dataset/src/main/scala/frameless/RecordEncoder.scala

```diff
@@ -44,10 +46,70 @@ object RecordEncoderFields {
     }
 }
 
-class RecordEncoder[F, G <: HList]
+trait NewInstanceExprs[T <: HList] extends Serializable {
+  def from(exprs: List[Expression]): Seq[Expression]
+}
+
+object NewInstanceExprs {
+
+  implicit def deriveHNil: NewInstanceExprs[HNil] = new NewInstanceExprs[HNil] {
+    def from(exprs: List[Expression]): Seq[Expression] = Nil
+  }
+
+  implicit def deriveUnit[K <: Symbol, T <: HList]
+    (implicit
+      tail: NewInstanceExprs[T]
+    ): NewInstanceExprs[FieldType[K, Unit] :: T] = new NewInstanceExprs[FieldType[K, Unit] :: T] {
+      def from(exprs: List[Expression]): Seq[Expression] =
+        Literal.fromObject(()) +: tail.from(exprs)
+    }
+
+  implicit def deriveNonUnit[K <: Symbol, V , T <: HList]
+    (implicit
+      notUnit: V =:!= Unit,
+      tail: NewInstanceExprs[T]
+    ): NewInstanceExprs[FieldType[K, V] :: T] = new NewInstanceExprs[FieldType[K, V] :: T] {
+      def from(exprs: List[Expression]): Seq[Expression] = exprs.head +: tail.from(exprs.tail)
+    }
+}
+
+trait DropUnitValues[L <: HList] extends DepFn1[L] with Serializable { type Out <: HList }
+
+object DropUnitValues {
+  def apply[L <: HList](implicit dropUnitValues: DropUnitValues[L]): Aux[L, dropUnitValues.Out] = dropUnitValues
+
+  type Aux[L <: HList, Out0 <: HList] = DropUnitValues[L] { type Out = Out0 }
+
+  implicit def deriveHNil[H]: Aux[HNil, HNil] = new DropUnitValues[HNil] {
+    type Out = HNil
+    def apply(l: HNil): Out = HNil
+  }
+
+  implicit def deriveUnit[K <: Symbol, T <: HList, OutT <: HList]
+    (implicit
+      dropUnitValues : DropUnitValues.Aux[T, OutT]
+    ): Aux[FieldType[K, Unit] :: T, OutT] = new DropUnitValues[FieldType[K, Unit] :: T] {
+      type Out = OutT
+      def apply(l : FieldType[K, Unit] :: T): Out = dropUnitValues(l.tail)
+    }
+
+  implicit def deriveNonUnit[K <: Symbol, V, T <: HList, OutH, OutT <: HList]
+    (implicit
+      nonUnit: V =:!= Unit,
+      dropUnitValues : DropUnitValues.Aux[T, OutT]
+    ): Aux[FieldType[K, V] :: T, FieldType[K, V] :: OutT] = new DropUnitValues[FieldType[K, V] :: T] {
+      type Out = FieldType[K, V] :: OutT
+      def apply(l : FieldType[K, V] :: T): Out = l.head :: dropUnitValues(l.tail)
+    }
+}
+
+class RecordEncoder[F, G <: HList, H <: HList]
   (implicit
     lgen: LabelledGeneric.Aux[F, G],
-    fields: Lazy[RecordEncoderFields[G]],
+    nonUnitFields: DropUnitValues.Aux[G, H],
```

Comment:
Could you change the name of your implicits to `i1`, `i2`, ..., `in`? The reasoning is that their time gives more info that their name, so having the types nicely aligned makes it more readable!